File: wsip/lab3/c/n_pierwsza.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>


typedef unsigned long num_t;

/**
 * @brief Podaj n-ta liczbe pierwsza
 * @brief n - numer liczby pierwszej (n >= 1)
 */
num_t n_prime(num_t n)
{
    if (n < 1)
        return 0;

    // Inicjalizacja tablicy
    num_t primes[n];
    memset(primes, 0, sizeof(num_t)*(n));
    primes[0] = 2;

    // Licznik liczb pierwszych w tablicy, liczba do sprawdzenia
    num_t prime_count = 1, num = 3;

    while (prime_count < n)
    {
        const num_t root = (num_t)sqrtf(num);
        int is_prime = 1;

        for (num_t i = 0; i < prime_count; i++)
        {
            if (primes[i] > root)
                break;
            
            if (num % primes[i] == 0)
            {
                is_prime = 0;
                break;
            }
        }

        // Jest pierwsza
        if (is_prime)
            primes[prime_count++] = num;
        
        num += 2;
    }
    
    return primes[n-1];
}


int main(int argc, char* argv[])
{
    int n = 0;
    printf("Podaj liczbe: ");
    scanf("%d", &n);
    printf("%d-ta liczba pierwsza: %lu\n", n, n_prime(n));
}

File: wsip/lab3/py/n_pierwsza.py
import numpy as np

def nth_prime(n: int) -> np.int64:
    """
    Oblicza n-tą liczbę pierwszą
    """

    if n < 1:
        raise ValueError(f'Niepoprawne dane, {n} < 1')

    primes      = np.zeros(n, dtype=np.int64)
    primes[0]   = 2
    prime_count = 1
    candidate   = 3

    while prime_count < n:
        is_prime = True

        for i in range(0, prime_count):
            # Sprawdzamy czy dzielnik jest wiekszy od pierwiastka z liczby
            # Wtedy nie ma sensu sprawdzanie dalszych dzielników
            if primes[i] ** 2 > candidate:
                break
            
            # Jeżeli liczba jest podzielna przez którąś 
            # z wcześniej znalezionych liczb pierwszych
            # to nie jest liczbą pierwszą
            if candidate % primes[i] == 0:
                is_prime = False
                break

        if is_prime:
            primes[prime_count] = candidate
            prime_count += 1

        candidate += 2

    return primes[n - 1]


def main():
    n: int = int(input('Podaj n: '))
    
    try:
        print(f'{n}-ta liczba pierwsza: {nth_prime(n)}')
    except ValueError as e:
        print(e)


if __name__ == '__main__':
    main()